frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.tsx

```diff
@@ -23,6 +27,22 @@ export function DashCardTabMenu({ dashCardId }: DashCardTabMenuProps) {
     return tabs.filter(t => t.id !== selectedTabId);
   }, [selectedTabId, tabs]);
 
+  const moveDashcard = useCallback(
+    (destTabId: number) => {
+      dispatch(moveDashCardToTab({ dashCardId, destTabId }));
+      dispatch(
+        addUndo({
+          message: t`Card moved to ${tabs.find(c => c.id === destTabId)?.name}`,
+          undo: true,
+          action: () => {
+            dispatch(undoMoveDashCardToTab({ dashCardId }));
+          },
+        }),
+      );
+    },
```

Comment:
`moveDashCardToTab` action seems to always have to call `addUndo` anyway. Should we call `addUndo` inside `moveDashCardToTab` instead then?

Reply:
Moved into a thunk, thanks!

Reply:
I'm not a huge fan of thunks because they distort the pure function style of redux and hide side effects. Without thunks you know that dispatching action X runs case X in the reducer, but with thunks any kind of side effects can happen. When thunks get too big and complex and start calling other thunks it becomes a nightmare to understand and maintain, like in [data-fetching.js]([url](https://github.com/metabase/metabase/blob/3398d6c4b85ae1fb343b3f71d82f32718cfc3bb6/frontend/src/metabase/dashboard/actions/data-fetching.js))

That's why originally with [`deleteTab`]([url](https://github.com/metabase/metabase/blob/3398d6c4b85ae1fb343b3f71d82f32718cfc3bb6/frontend/src/metabase/dashboard/components/DashboardTabs/use-dashboard-tabs.ts#L32)) I dispatched the `addUndo` separately, to make it explicit at the call site that there's a side effect.

To be fair though this use case isn't egregious so I don't think it's necessary to refactor, but it would be nice if we could avoid thunks more often going forward in favor of explicitly dispatching separate actions at the callsite (which itself can be a reusable function that isn't a redux action).

Reply:
Has there been any discussion in our FE team regarding thunks since I'm not aware of it? This could be a good RFC if we haven't discussed it yet.

I agree `data-fetching.js` isn't a very good example when it comes to using thunks. But I was thinking about the relationship between actions. At a glance, if calling `deleteTab` or `moveTab` means we also need to call `addUndo` then I don't see the point of repeating `addUndo` in every call site.

However, that could become problematic if we ever decided to not show undo toast in certain places. And I do like code that looks consistent, knowing how you've done with `deleteTab`

I think calling `moveDashCardToTab` and `addUndo` side by side is better in this regard.

@npretto can you bring the original change back?